tests/simulators/carla/test_actions.py

```diff
@@ -0,0 +1,133 @@
+import os
+from pathlib import Path
+import signal
+import socket
+import subprocess
+import time
+
+import pytest
+
+try:
+    import carla
+
+    from scenic.simulators.carla import CarlaSimulator
+except ModuleNotFoundError:
+    pytest.skip("carla package not installed", allow_module_level=True)
+
+from tests.utils import compileScenic, sampleScene
+
+
+def checkCarlaPath():
+    CARLA_ROOT = os.environ.get("CARLA_ROOT")
+    if not CARLA_ROOT:
+        pytest.skip("CARLA_ROOT env variable not set.")
+    return CARLA_ROOT
+
+
+def launchCarlaServer():
+    CARLA_ROOT = checkCarlaPath()
+    subprocess.Popen(f"bash {CARLA_ROOT}/CarlaUE4.sh -RenderOffScreen", shell=True)
+
+    for _ in range(30):
+        if not isCarlaServerRunning():
+            time.sleep(1)
+
+    # extra 3 seconds to ensure server startup
+    time.sleep(3)
+    return
+
+
+def terminateCarlaServer():
+    result = subprocess.run("lsof -i :2000", capture_output=True, text=True, shell=True)
```

Comment:
This looks somewhat risky and fragile. How about just having `launchCarlaServer` return the `Popen` object and then calling `terminate()` on that here? It looks like you're supporting having CARLA already running when `pytest` starts but I'm not sure if we want to do that given that we're unconditionally killing CARLA here. It's probably better to have our fixture completely manage CARLA, both starting and stopping it (in which case there may be no need to have these helper functions and you could just inline them into the fixture).

Reply:
Since the CARLA server is launched from another bash script we need to trigger `killall` for relevant processes. But coupling starting and stopping into the fixture has been performed!